go/worker/compute/executor/committee/node.go

```diff
@@ -534,15 +539,17 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
 				"header_hash", header.EncodedHash(),
 				"header_type", header.HeaderType,
 			)
-			if header.HeaderType != block.Normal {
-				return
-			}
-			if !header.IORoot.Equal(&state.proposedIORoot) {
+
+			if header.HeaderType != block.Normal ||
+				!header.IORoot.Equal(&state.proposedIORoot) {
 				n.logger.Error("proposed batch was not finalized",
 					"header_io_root", header.IORoot,
 					"proposed_io_root", state.proposedIORoot,
+					"header_type", header.HeaderType,
 					"batch", state.raw,
 				)
+				// Reinsert the failed batch.
+				n.reinsertTxs(state.raw)
```

Comment:
Hm, aren't the processed transactions only removed below (line 564)? Where have they been removed in this case so they need to be reinserted?

Reply:
In case we are the "proposer" (scheduler), the txs are removed at `Dispatch`, but actually a good point, if we switch the logic to only remove the transactions after a successful round finalization (and not at dispatch time) we can probably make the code a lot simpler.

Will look into doing that